Modules/grpmodule.c

```diff
@@ -119,16 +120,47 @@ grp_getgrgid_impl(PyObject *module, PyObject *id)
         }
         Py_DECREF(py_int_id);
     }
+#ifdef HAVE_GETGRGID_R
+    Py_BEGIN_ALLOW_THREADS
+    int status, bufsize = NSS_BUFLEN_GROUP;
+    struct group grp;
 
-    if ((p = getgrgid(gid)) == NULL) {
+    buf = PyMem_RawMalloc(bufsize);
+
+    do {
+        status = getgrgid_r(gid, &grp, buf, bufsize, &p);
+        if (p != NULL || status != ERANGE) {
+            break;
+        }
+        if ((bufsize << 1) > (NSS_BUFLEN_GROUP << 10)) {
+            status = ERANGE;
+            break;
+        }
+        bufsize <<= 1;
+        buf = PyMem_RawRealloc(buf, bufsize);
```

Comment:
`PyMem_RawRealloc()` can fail. In this case it will return NULL.
```c
p = PyMem_RawRealloc(buf, bufsize);
if (p == NULL) {
    PyErr_NoMemory();
    break;
}
buf = p;
```